crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+    /// algorithm will not work properly and is very likely to panic.
+    pub(super) unsafe fn new(
+        updates: Arc<RwLock<UpdatesInner<Item, Gap>>>,
+        token: ReaderToken,
+        initial_chunk_lengths: VecDeque<(ChunkIdentifier, ChunkLength)>,
+    ) -> Self {
+        Self { updates, token, mapper: UpdateToVectorDiff { chunks: initial_chunk_lengths } }
+    }
+
+    /// Take the new updates as [`VectorDiff`].
+    ///
+    /// It returns an empty `Vec` if there is no new `VectorDiff` for the
+    /// moment.
+    pub(super) fn take(&mut self) -> Vec<VectorDiff<Item>>
+    where
+        Item: Clone,
+    {
+        let mut updates = self.updates.write().unwrap();
+        let updates = updates.take_with_token(self.token);
+
+        self.mapper.map(updates)
+    }
+}
+
+/// Internal type that converts [`Update`] into [`VectorDiff`].
+///
+/// This type maintains its own state.
+struct UpdateToVectorDiff {
+    /// Pairs of all known chunks and their respective length. This is the only
+    /// required data for this algorithm.
+    chunks: VecDeque<(ChunkIdentifier, ChunkLength)>,
+}
+
+impl UpdateToVectorDiff {
+    /// Map several [`Update`] into [`VectorDiff`].
+    ///
+    /// How this type transforms `Update` into `VectorDiff`? There is no
+    /// internal buffer of kind [`eyeball_im::ObservableVector<Item>`],
+    /// which could have been used to generate the `VectorDiff`s. They are
+    /// computed manually.
+    ///
+    /// The only buffered data is pairs of [`ChunkIdentifier`] and
+    /// [`ChunkLength`]. The following rules must be respected:
+    ///
+    /// * A chunk of kind [`ChunkContent::Gap`] has a length of 0,
+    /// * A chunk of kind [`ChunkContent::Items`] has a length equals to its
+    ///   number of items,
+    /// * The pairs must be ordered exactly like the chunks in [`LinkedChunk`],
+    ///   i.e. the first pair must represent the first chunk, the last pair must
+    ///   represent the last chunk.
+    ///
+    /// The only thing this algorithm does is maintaining the pairs:
+    ///
+    /// * [`Update::NewItemsChunk`] and [`Update::NewGapChunk`] are inserting a
+    ///   new pair with a chunk length of 0 at the appropriate index,
+    /// * [`Update::RemoveChunk`] is removing a pair,
+    /// * [`Update::PushItems`] is increasing the length of the appropriate pair
+    ///   by the number of new items, and is potentially emitting
+    ///   [`VectorDiff`],
+    /// * [`Update::DetachLastItems`] is decreasing the length of the
+    ///   appropriate pair by the number of items to be detached; no
+    ///   [`VectorDiff`] is emitted,
+    /// * [`Update::ReattachItems`] and [`Update::ReattachItemsDone`] are
+    ///   respectively muting or unmuting the emission of [`VectorDiff`] by
+    ///   [`Update::PushItems`].
+    ///
+    /// The only `VectorDiff` that are emitted are [`VectorDiff::Insert`] or
+    /// [`VectorDiff::Append`] because a [`LinkedChunk`] is append-only.
```

Comment:
We should also be able to `Clear` a full linked chunk, in case we're seeing a limited sync (until we have a fancier diff/reconciliation algorithm)?

Reply:
That's something I want to discuss with you. `RoomEventCacheUpdate` has a `Clear` variant already. Clearing a `LinkedChunk` means removing data from the memory and probably the database. Clearing a `RoomEventCache` means removing data from the memory only as far as I know. The meaning and the usage aren't exactly the same. Let's discuss about that and maybe let's open a new PR if necessary.

Reply:
Also, to have a `VectorDiff::Clear` here, we must first have an `Update::Clear`, which isn't the case so far. Let's handle that as a separate feature patch.